### DISCUSSION TITLE
A guide on best practices for developing Python models in dbt 🐍

### DISCUSSION THREAD

**Discussion by joellabes** [dbt-labs/docs.getdbt.com#1811] [Archive]
As [Python model support in dbt](https://github.com/dbt-labs/dbt-core/discussions/5261) comes closer, it's time to start thinking about guides to help practitioners do things in a dbtonic way. 

There are two directions we can view this from; I think they're probably different enough that we'll need a guide for each persona, but maybe they can overlap! 
1. The dbt user who wants to do some modelling in Python, so that they can do things SQL isn't good at but stay inside of their familiar context
2. The Python user who wants to start hosting their projects in dbt, so that they can take advantage of dbt's existing data models and understanding of lineage instead of babysitting a stack of notebooks. 

To reiterate @jtcohen6's callout in the above discussion: 
>Our primary intent with Python models is to extend dbt’s reach into solving new types of problems, by opening up new kinds of modeling to people for whom it was previously inaccessible: people for whom the data warehouse is their principal tool.

Fortunately, I represent persona # 1 pretty well! Here’s some things I'm hoping we can build out: 

- Several worked examples that broaden my horizons: I sort of conceptually understand that Python is cool and you can do predictive stuff with it, but don't know where to start. 
- Flagging common pitfalls: how might [thinking with SQL](https://youtu.be/TluRVBhmf8w?t=135) get in my way when I start to develop models in Python? 
- ⚠️ I don't want us to build _here's how you pivot a table with dbt_utils, and here's how you pivot a table with pandas_, since our goal is to unlock net-new capabilities without unnecessarily stripping SQL out of the mix 
    - which I guess adds another thing to the list: some sort of decision tree on what stuff even needs to be in Python
- But I do want to be able to bootstrap a mental model based on what I already know about Analytics Engineering. What is the equivalent of a dbt macro in Python? How should I implement environment-aware filtering? 
- How and where should people develop their models before adding them to the dbt project? 

@lostmygithubaccount will represent the dbt-curious Python developer, but for sure we'll want some the same sort of pitfall guidance: what behaviours make sense in other Python contexts but don't make sense in a dbt world?

**Comment by xg1990** [comment]
Curious is it the right link to [thinking with SQL](https://youtu.be/TluRVBhmf8w?t=135)?

**Reply by joellabes** [reply]
Heh, yep it's just a very bad joke referencing the "now you're thinking with portals" quote.

**Reply by Opiano1** [reply]
I had to rewatch the url again to confirm😂 and seeing it was meant to be a bad joke

**Comment by kenfar** [comment]
One of the most valuable features of using python for transformations that was not covered in the documentation is the ability write automated unit tests.  Not quality-control checks against existing data, but quality-assurance checks against potential data.  To test numeric overflow, invalid dates, encoding issues, validation & exception logic, etc, etc.   I would really love to see how folks would leverage this capability:  move each field transform into a dedicated python function and then accompany that with a unit test class?

**Reply by lostmygithubaccount** [reply]
hi @kenfar, we're absolutely thinking about how to support that! in v1.3 however, testing is still done through the existing generic and singular tests authored in SQL. this is in our list of next steps on discovery work (see this discussion: https://github.com/dbt-labs/dbt-core/discussions/5742)

**Reply by AndrewLane** [reply]
I've been trying to explore how I'd do this "unit" testing (where the tests still reach out to Snowflake for computation) given the tools available today, and this is what I've come up with.  Would appreciate any criticism or comments:

https://github.com/AndrewLane/dbt-python-model-testing-exploration

**Reply by lostmygithubaccount** [reply]
very cool! we'll definitely be taking a look 👀 thanks for developing this @AndrewLane!

**Reply by bbrewington** [reply]
I just stumbled on this discussion in 2026, so thought I'd drop this gem in thread: https://docs.getdbt.com/docs/build/unit-tests
